examples/src/main/java/kafka/examples/Producer.java

```diff
@@ -21,133 +21,164 @@
-        String messageStr = "Message_" + messageKey;
-
-        if (isAsync) { // Send asynchronously
-            sendAsync(messageKey, messageStr, currentTimeMs);
-            return;
+    public void shutdown() {
+        if (!closed) {
+            closed = true;
+            latch.countDown();
         }
-        Future<RecordMetadata> future = send(messageKey, messageStr);
-        future.get();
-        System.out.println("Sent message: (" + messageKey + ", " + messageStr + ")");
     }
 
-    private void sendAsync(final int messageKey, final String messageStr, final long currentTimeMs) {
-        this.producer.send(new ProducerRecord<>(topic,
-                        messageKey,
-                        messageStr),
-                new DemoCallBack(currentTimeMs, messageKey, messageStr));
+    public KafkaProducer<Integer, String> createKafkaProducer() {
+        Properties props = new Properties();
+        // bootstrap server config is required for producer to connect to brokers
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+        // client id is not required, but it's good to track the source of requests beyond just ip/port
+        // by allowing a logical application name to be included in server-side request logging
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "client-" + UUID.randomUUID());
+        // key and value are just byte arrays, so we need to set appropriate serializers
+        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        if (transactionTimeoutMs > 0) {
+            // max time before the transaction coordinator proactively aborts the ongoing transaction
+            props.put(ProducerConfig.TRANSACTION_TIMEOUT_CONFIG, transactionTimeoutMs);
+        }
+        if (transactionalId != null) {
+            // the transactional id must be static and unique
+            // it is used to identify the same producer instance across process restarts
+            props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId);
+        }
+        // enable duplicates protection at the partition level
+        props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, enableIdempotency);
+        return new KafkaProducer<>(props);
     }
 
-    private Future<RecordMetadata> send(final int messageKey, final String messageStr) {
-        return producer.send(new ProducerRecord<>(topic,
-                messageKey,
-                messageStr));
+    private void asyncSend(KafkaProducer<Integer, String> producer, int key, String value) {
+        // send the record asynchronously, setting a callback to be notified of the result
+        // note tha send blocks when buffer.memory is full or metadata are not available
```

Comment:
Sorry, this is not correct. The record will be sent when either `batch.size` or `linger.ms` reached. Please refer to [doc](https://kafka.apache.org/documentation/#producerconfigs_batch.size). I don't think buffer.memory will affect the send. Am I correct?

Reply:
There is a typo, but I think it is correct. 

If a producer needs to refresh the metadata and they are not available it will wait for them, blocking the send() operation and raising a warning. 

If messages are sent faster than they can be delivered to brokers, the producer may run out of buffer.memory and the next send() call will either block or throw exception, depending on max.block.ms. Try to set buffer.memory==batch.size and then check bufferpool-wait-ratio producer metric, you will see that it's very close to or at 100%, meaning that send() is always waiting on previously sent requests.

Reply:
OK, I know what you mean. But `batch.size` and `linger.ms` can also block the send, why do we only mention `buffer.memory` and metadata unavailable here? Let's say, we set a large number of `batch.size` and `linger.ms`, that means we won't send any data until one batch is full or linger.ms expired, is that right? 

I'm thinking we don't need to write them all here. We can link them to the producer javadoc like this:
```
note that send will be blocked when buffer.memory is full or metadata not available or other reasons. See {@link KafkaProducer} for more info.
```
WDYT?

Reply:
Yeah, I was only mentioning buffer.memory because it is the less obvious of the three. What about "note that even if you set a small batch.size will linger.ms=0, the send operation will still be blocked when buffer.memory is full or metadata are not available"?

Reply:
Sounds good.
nit: note that, even if you set a small batch.size [with] linger.ms=0, ...